src/v/cluster/rm_stm.cc

```diff
@@ -469,6 +475,11 @@ ss::future<tx_errc> rm_stm::do_prepare_tx(
     co_return tx_errc::none;
 }
 
+void rm_stm::on_leadership_change(
```

Comment:
We have two source of where the modifications of rm_stm may come from:
  - rpc requests from kafka clients or txn coordinator
  - from the apply caused by replaying the log

`on_leadership_change` adds a third source of the changes and it forces us to think about more of combinations. For example, can a request from a client of a txn coordinator come before on_leadership_change is triggered, what happens if on_leadership_change fails.

Let's use sync to avoid adding a new entity. Sync guarantees that all records written by previous leader are already applied. So an apply can process `send_tx_leadership` & `receive_tx_leadership` as it goes. When sync sees unbalanced `send_tx_leadership` it should write `receive_tx_leadership` before yielding control flow. Also several syncs may be running in parallel so once it sees unbalanced `send_tx_leadership` it should use mutex to guarantee that only one `receive_tx_leadership` is written